### ISSUE TITLE
Inconsistent Pattern Generating Table_of_Symbol?

### ISSUE THREAD

**Issue by halonazhao** [JacquesCarette/Drasil#555]
SSP, NoPCM and SWHS are using tsymb'' which leads to use SI{_concepts}; 
GlassBR, GamePhysics are using tsymb which leads to use SI{_quants};
This doesn't make sense because generating the same reftab is suppose to use the same field of the SI information if different fields have quite independent meaning.

https://github.com/JacquesCarette/Drasil/blob/97069a74c7ee81c7cf5c4bcbb76cffd5608089a1/code/Example/Drasil/DocumentLanguage.hs#L245-L290

But Dan said we are finally going to use field "_sysinfodb" which is the at ChunkDB type. Just want to make sure we will finally use "_sysinfodb"?  After we have the new templet, I will adjust the unit collection to match. I will start doing the symbol collection first.

@szymczdm @JacquesCarette

**Comment by JacquesCarette** [comment]
Another excellent question!  I think what this means is that there is currently two different styles of tables of units. That they are grabbing their information from different places is indeed inconsistent. However, the `SI{_quants}` and `SI{_concepts}` fields are not all that different from each other, so "quite independent meaning" is inaccurate.  We need to understand if this choice is important - so perhaps both @szymczdm and @smiths can look at this.

Dan is right that we are going to use the `ChunkDB` part of the system information as the better way to collect this information.

As an experiment, what happens if you switch SSP, NoPCM, and SWHS to use `tsymb` ? And what about the opposite, GlassBR, GamePhysics to `tsymb''` ?  That could help us better understand what these choices really mean.

**Comment by halonazhao** [comment]
The key difference here is that `tsymb` cannot support choosing display from `Defn`, `Term`, `TermExcept` and `TermAndDefn`.

But `tsymb''` can support choosing which field we want to generate.

So it doesn't make sense apply `tsymb` to SSP, NoPCM and SWHS because they need choosing field.
If apply`tsymb''` to GlassBR and GamePhyscis , it would be like below:

The calling function would be changed from`tsymb intro` to `tsymb'' intro Term` where 'Term' is specified for choosing field.

**Comment by JacquesCarette** [comment]
That makes sense.  Ok, the right thing to do then is

1. change GlassBR and GamePhysics to use `tsymb'' intro Term`. Check that in and make a PR.
2. remove `tsymb` (and the part of the `DocumentLanguage` that leads there. Make sure all the examples still work. Check in and PR.
3. rename `tsymb''` to `tsymb`. Fix the examples. Check in and PR.

It makes sense to make the 'choices' more explicit. And the best workflow is to break things down into very small tasks, each of which go from "everything works" to "everything works".

**Comment by halonazhao** [comment]
I just found another problem. `tsymb''` uses `SI{_concepts}` and `concepts` has type `[DefinedQuantityDict]`. `tsymb` uses `SI{_quants}` and `quants` has type `[QuantityDict]`.

Because they are using different databases, so all the Unitals files wrap those conpects to different types, `[DefinedQuantityDict]` or `[QuantityDict]`. Still that question, do the actual meanings of `concepts` and `quants` have differences? Do we keep both database?

**Comment by JacquesCarette** [comment]
They have very similar meanings, because they share almost all of their fields. We should definitely keep both, but perhaps the way in which it is implemented is not great. Need to think about that.

But, for grabbing the units, I don't think this makes any difference!

**Comment by halonazhao** [comment]
Yes, they have nothing to do with the 'Unit' grabbing. Symbols are complete whichever database we use.

**Comment by szymczdm** [comment]
`tsymb''` using `[DefinedQuantityDict]` is an old (not good) design, when I mentioned the need for using `ChunkDB`, we should no longer require a specific type of chunk, only something that has a `uid`. From there, we should be able to build the table by looking up the appropriate fields (term or defn, units, etc) in the ChunkDB.

**Comment by JacquesCarette** [comment]
So @szymczdm what should @halonazhao do next?  Seems my plan was not quite right.

**Comment by szymczdm** [comment]
I had a chat with @halonazhao after our meeting today. The main issue (now) it seems is getting all of the appropriate units' information after parsing through the symbols (@halonazhao can correct me if I'm misquoting), particularly when a given unit occurs only within a more complex, derived unit.

Ex:

Given | Extracted
------|----------
m<sup>2</sup> | m
J | J
W/m<sup>2</sup> | 

So the table of units ends up lacking W.

I believe the approach to use would be to parse through every unit definition and grab every unit symbol which appears, then look up requisite information for this (`nub`ed) list in the SI units and generate the table. Essentially, it's just another database lookup, but from a database contained in Data.Drasil

**Comment by JacquesCarette** [comment]
I was afraid that would happen.  What this means is that, units, like other things, are too display-oriented. The derived units should be composed of other *units*, not merely other units' **symbols**.  That would obviate the need to reverse-lookup of unit symbols, as all we'd ever see are unit `uid`s. Which would likely simplify things quite a bit, as the current combinators for forming units are sub-optimal the way they currently work partly on units and partly on symbols-for-units.

**Comment by JacquesCarette** [comment]
If by "the units themselves" you mean "the units' uids", then yes, I agree, that's the way to go.

**Comment by halonazhao** [comment]
The key problem of getting the fundamental units out from a derived one is about the way we build unit equations. Building equation actually lose a lot of information, because it equation takes `UnitDefn` but  directly produce a `USymb`.

Whatever how we play with the `UDefn` will just help us distinguish fund unit and derived unit, but not parse the fund unit from derived ones.

I think I need better way to build unit equation so that it doesn't lose information. I am working on that now. 

If I went to a wrong direction, please correct me.

**Comment by JacquesCarette** [comment]
You have analyzed the problem correctly. You are going in the correct direction. I would like to see some details of your design before you implement a lot of code. I want to avoid wasting your time implementing something that "won't work", if I can predict it ahead of time.

**Comment by halonazhao** [comment]
Ok. Here is my thinking:

Starting from building the unit equation, first, I think I need one new data type `SymbEquation` which has field `{_contributingUnit :: UnitDefn, _usymb :: USymb}`. I think why we need a new data type instead of directly using `UnitDefn` is that when the new unit defining is in the middle like `newton = kilogram *$ (second ^: (-2))`, it needs several operations to finalize it. But in the middle of building it like `(second ^: (-2))`, it is not an actual `UnitDefn` becasue it has no actual meaning.

Being specific to build a SymbEquation, there should be two different groups. First group would include, for example, `(^:) :: IsUnit -> u -> Integer -> USymb`,  which take `UnitDefns` to produce a `USymb`. I would change constructors in this category to `(^:) :: IsUnit -> u -> Integer -> SymbEquation `.

Another category would be like `(*$) :: (IsUnit u1) => u1 -> USymb -> USymb`, which take new SI unit and a symbol to produce new symbol. I want to change the type to `(*$) :: (IsUnit u1) => u1 -> SymbEquation -> SymbEquation`. When it is producing the new the SymbEquation, it automatically taking the `contributingUnits` information from the input SymbEquation.

So both groups of constructors would collect all contributing units information completely.

But so far, all of these only can make sure that information didn't get lost, another problem would be going up to the new unit building level which actually use this new `SymbEquation` structure.

The whole old procedure to build a new derived unit is , for example:
`s_2 :: UnitDefn`
`s_2 = new_unit "seconds squared" $ second ^: 2`
`new_unit :: String -> USymb -> UnitDefn`
`new_unit s u = makeDerU (unitCon s) (USynonym u)`
`makeDerU :: ConceptChunk -> UDefn -> UnitDefn`
`makeDerU concept eqn = UD concept (from_udefn eqn) (Just eqn)`

It seems very hard to parse out anything directly from a `UnitDefn` type variable(`s_2 :: UnitDefn`). This part I'm not sure. I looked through some pattern matching examples, but I didn't find similar case.

So my thinking is add an extra field [UnitDefn] to record the contributing units in data type `UnitDefn`. 
I talked with Dan the other day, I told him about the idea adding [UnitDefn] field to `UnitDefn`. He said that would be more burden to user to manually insert all contributing unit information to a unit, which I agree at that point.

But after I have new data type `SymbEquation` which accumulate all the contributing units, I think we could have smarter constructor to get the information from `SymbEquation` to automatically fill the `[UnitDefn]` field. User as before only type equation in, and the smarter constructor retrieving the information and give to `[UnitDefn]`. For the fundamental units, we can also use smarter constructor to detect `Maybe Udefn` field which can help to recognize the fundatmental units, and the constructor automatically puting nothing into `[UnitDefn]` field.

After all of these, that would be very easy to get any information out from `UnitDefn`.

I don't know whether this way would work and it seems like a complicate way to achieve this.

**Comment by halonazhao** [comment]
This way is really messy. Even though I can think through once, I would forget again. I think it's better I kind of giving a demo to make myself clear.

I will take `s_2 :: UnitDefn` 
     `s_2 = new_unit "seconds squared" $ second ^: 2` as an example. In this case the equation at `SymbEquation` type is `second ^: 2`

`new_unit1 :: String -> SymbEquation -> UnitDefn`
`new_unit1 s se = makeDerU (unitCon s) se`
`makeDerU :: ConceptChunk -> SymbEquation -> UDefn -> UnitDefn`
`makeDerU concept se = UD concept (from_udefn (USynonym (se ^. usymb))) (Just (USynonym (se `
`^. usymb))) (view [UnitDefn] of se)`     ----The last element has wrong syntax but mean it

**Comment by JacquesCarette** [comment]
First, our current combinators might be "somewhat wrong". We want to have combinators which make it easy to create new units from old -- all the rest of the details can change. Right now, some cases of "new units from old" are easy, some are still hard.  So that means the design is flawed.

What you propose is a step forward, but I don't think it is enough.  I think the real change needs to be in `symbEquation`: it is not just the units that make up the current one that need to be remembered, but the whole 'equation'.  meter/second is different than meter*second in many different ways.

So I think that `USymb` should change to have the `uid` as well as the `Symbol` of the units used to create it.

**Comment by JacquesCarette** [comment]
I meant
```haskell
newtype USymb = US [(UID, Symbol, Integer)]
```
but even that isn't right either!  Let me think about it a little more.

**Comment by halonazhao** [comment]
Ok. Let me rephrase my understanding. So we have three types of units we have to distinguish from each other. 
`data UnitDefn = UD { _vc :: ConceptChunk, _u :: USymb, _ud :: Maybe UDefn}`

First type is like `meter` which is a basic SI unit, when we build this unit we have an `uid` attach to this unit which is retrieved from `_vc :: ConcecptChunk`. You want the `_u :: USymb` to have a same uid(I am not sure whether we need a different one). But the uid for USymb definitely help us to display a symbol by finding uid.

The second type is like 'joule', which is a derived SI unit. Although the symbol for it should be `J`, `kilogram*m^2*secod^(-2)` is also recorded as another style of symbol.

The third type is like `joule / centigrade`, a pure derived unit. The symbol for this is the result from `USymb` field in our `SymbEquation`. Because it is a new unit, it will have ConceptChunk for it, so the uid for `USymb`  can be the uid for this unit as well.
  
`newtype USymb = US [(UID, Symbol, Integer)]` will work fine for the first and the third type of unit.

The only problem would be the second type because it has two symbols. So my thought would be change the UnitDefn type to `data UnitDefn = UD { _vc :: ConceptChunk, _u :: [USymb], _ud :: Maybe UDefn}` and change the pattern of USymb like `short USymb` and `long Usymb`. But give this two symbols the same uid, when we want to retrieve a specific one, we can pattern filter out the needed one. I think after optimizing this structure, the `_ud :: Maybe UDefn` could be removed as well.

Also, if this understanding is relatively right, I think the uid for unit and the uid for the symbol of unit can be the same. So we don't have to give symbol another uid. Even we use equation to build new unit, the generated symbol(at USymb type) will be fully give back to UnitDefn. It didn't lose information. I mean even meter*second or meter/second are produced to symbol and give back to the new unit(UnitDefn) which has uid itself.

**Comment by halonazhao** [comment]
I just implemented my earlier thought to code. And it is now working, but so far only implemented it for SWHS case. But it still easy to check and trace down. I know you didn't prove my design yet, but just for practice. After you have better thoughts about this, I am happy to make any changes.

**Comment by JacquesCarette** [comment]
I finally got some time to come back to this. I really like your analysis above. It made the issues really clear.

There are two things that I do not really like about the resulting design though:
1. that you have a list of `USymb`. I think you nicely justify that we could have either 1 or 2 different symbols for a unit.  But with a list, you could also have 0 or 23, both of which are harder to justify. I think we might want 2 fields (one `USymb` and one `Maybe USymb`) where the first would be the symbol for the unit in terms of fundamental units only, and the second would be for display (such as for joule).
2. that for pure derived units, it is not possible to know how the unit was constructed. This is a problem with `SymbEquation` (I think), which should be in terms of other units' `UID` rather than in terms of their symbols.

**Comment by halonazhao** [comment]
So far I define UnitEquation as `data UnitEquation = UE {_contributingUnit :: [UnitDefn], _us :: USymb}`. U want me change `[UnitDefn]` to `[UID]` instead?

**Comment by JacquesCarette** [comment]
Yes.

**Comment by halonazhao** [comment]
I just pushed the new version https://github.com/JacquesCarette/Drasil/commit/68c84d36f78e1c72edeba85bad4081a9026788f1. Currently the `UnitDefn = UD { _vc :: ConceptChunk, 
                     _fsymb :: USymb,
                     _dsymb :: Maybe USymb,
                     _ud :: Maybe UDefn,
                     _cu :: [UID]}` and `data UnitEquation = UE {_contributingUnit :: [UID], _us :: USymb}`.

There is a problem. The `_cu` field in UnitDefn and `_contributing` field in UnitEquation kind of have to be consistent. We are using UnitEquation to generate new unit and store the contributing unit in `_cu`. So I put the [UID] for `_cu`. If at this level, it is hard to use UID to get back to a `Chunk` as well because database for a specific example is invisible here. 

But this would cause confusion because not only unit have UID, other chunks may have UID as well. Also, once we directly collect UID, when we getting back to chunk at Example level, the `getUnitLup :: HasSymbolTable s => (HasUID c) => c -> s -> Maybe UnitDefn` functions doesn't work. We have to have a new function directly take UID and get chunk, like `getUnitLup' :: HasSymbolTable s => UID -> s -> Maybe UnitDefn`.

**Comment by JacquesCarette** [comment]
You are absolutely right that those two fields should be consistent.

Yes, it is indeed potentially confusing that there are multiple `UID` around.  I am not sure that taking in a `UID` directly in `getUnitLup`' necessarily makes things better.  Somehow we should know that the first parameter to our function should refer to a unit.  Perhaps a `HasUnit` constraint?

**Comment by halonazhao** [comment]
After thinking about the constraints, `HasUnit` which means `Unitary` doesn't really apply here. Since we will give `DefinedQuantityDict` to `getUnitLup`, but `DefinedQuantityDict` only have a `Maybe UnitDefn` field. It doesn't always have a unit, after thinking about this, I think `Quantity` and `HasUID` would be proper Constraints. So it finally came out as `getUnitLup :: HasSymbolTable s => (Quantity c, HasUID c) => c-> s -> Maybe UnitDefn`. Is `Quantity` proper here or just like the old one only having `HasUID` constraint?

**Comment by JacquesCarette** [comment]
Can you recap the issue?  The conversation above is very long, and I am no longer sure that I understand what the remaining problems are.  So I am afraid to give advice without properly understanding the issue.

**Comment by halonazhao** [comment]
Ok. There is only one problem left which is choosing the most proper constraints for `getUnitLup` function in ChunkDB.hs.

This function was defined as `getUnitLup :: HasSymbolTable s => (HasUID c) => c -> s -> Maybe UnitDefn`, which indicates that we may get a Unit from a chunk who has UID.

As you said `Somehow we should know that the first parameter to our function should refer to a unit`, if only give a constraint "HasUID" make the input range too wide, I was thinking about constraint "Quantity", because this is the closest one to possibly getting a unit. `Unitary` is most decent conceptual constraint for `refer a unit`, but this function's input set includes some variables without units. So if we step back, `Quantity` is the most strict constraint for this function.

**Comment by JacquesCarette** [comment]
Don't we have `MayHaveUnit` ?  That would be perfect.

**Comment by JacquesCarette** [comment]
Otherwise `Quantity` is 2nd best.

**Comment by JacquesCarette** [comment]
I took a long time answering because I was not sure what best to do here.  Let's go with `MayHaveUnit`, with a `unitOpt` member function.

**Comment by JacquesCarette** [comment]
Yes, that's ok.  No, `MayHaveUnit` should not inherit from `Quantity`.  Instead `Quantity` should become
```haskell
class (Idea c, HasSpace c, HasSymbol c, MayHaveUnit) => Quantity c where
```

Of course that change will require ~20 files or so to be changed, but that should be straightforward.

Thanks for thinking this through. The end results will be much nicer than if you'd just implemented my too-simple suggestion.

**Comment by halonazhao** [comment]
This issue can be closed now since collecting units branch has been merged into Master.
